network/simple-onchain-discovery/src/lib.rs

```diff
@@ -68,16 +77,112 @@ pub static NETWORK_KEY_MISMATCH: Lazy<IntGaugeVec> = Lazy::new(|| {
+    ValidatorSet(ValidatorSetChangeListener),
+    File(FileChangeListener),
+}
+
+impl DiscoveryChangeListener {
+    pub fn start(self, executor: &Handle) {
+        match self {
+            Self::ValidatorSet(listener) => executor.spawn(Box::pin(listener).start()),
+            Self::File(listener) => executor.spawn(Box::pin(listener).start()),
+        };
+    }
+
+    pub fn discovery_source(&self) -> DiscoverySource {
+        match self {
+            DiscoveryChangeListener::ValidatorSet(listener) => listener.discovery_source(),
+            DiscoveryChangeListener::File(listener) => listener.discovery_source(),
+        }
+    }
+}
+
+/// A listener for changes on an incoming discovery source.  The entire set of discovered peers must
+/// be provided in each update.
+#[async_trait]
+pub trait DiscoveryChangeListenerTrait: Stream<Item = PeerSet> {
+    /// `NetworkContext` for logging
+    fn network_context(&self) -> &NetworkContext;
+
+    /// The `DiscoverySource` for purposes of keeping track of each source separately
+    fn discovery_source(&self) -> DiscoverySource;
+
+    /// The `channel::Sender` for updating networking with a new `PeerSet`
+    fn update_channel(&self) -> &Mutex<channel::Sender<ConnectivityRequest>>;
+
+    async fn start(mut self: Pin<Box<Self>>) {
+        let discovery_source = self.discovery_source();
+        info!(
+            NetworkSchema::new(self.network_context()),
+            "{} Starting {} Discovery",
+            self.network_context(),
+            discovery_source
+        );
+
+        while let Some(update) = self.next().await {
+            trace!(
+                NetworkSchema::new(self.network_context()),
+                "{} Sending update: {:?}",
+                self.network_context(),
+                update
+            );
+            let request = ConnectivityRequest::UpdateDiscoveredPeers(discovery_source, update);
+            if let Err(error) = self.update_channel().lock().try_send(request) {
+                inc_by_with_context(&DISCOVERY_COUNTS, self.network_context(), "send_failure", 1);
+                warn!(
+                    NetworkSchema::new(self.network_context()),
+                    "{} Failed to send update {:?}",
+                    self.network_context(),
+                    error
+                );
+            }
+        }
+        warn!(
+            NetworkSchema::new(&self.network_context()),
+            "{} {} Discovery actor terminated",
+            self.network_context(),
+            self.discovery_source()
+        );
+    }
+}
+
 /// Listener which converts published  updates from the OnChainConfig to ConnectivityRequests
 /// for the ConnectivityManager.
 pub struct ValidatorSetChangeListener {
     network_context: Arc<NetworkContext>,
     expected_pubkey: PublicKey,
     encryptor: Encryptor<Storage>,
-    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    conn_mgr_reqs_tx: Mutex<channel::Sender<ConnectivityRequest>>,
```

Comment:
wow, really? I guess this one isn't capable of being shared?

Reply:
We could clone it, I just needed to get a mutable object from an immutable object

Reply:
you can clone senders, don't need the mutex : )